empress/recon_vis/recon_viewer.py

```diff
@@ -96,83 +222,358 @@ def render_parasite_helper(fig, node, recon, host_lookup, show_internal_labels,
+    :param recon: Reconciliation object
+    :param fig: Figure object that visualizes trees using MatplotLib
+    :param stop_row: row number to stop line drawing on
+    """
+    mapping_node = recon.mapping_of(child_node.name)
+    host_node = host_lookup[mapping_node.host]
+    
+    if stop_row == None:
+        stop_row = node.layout.row
+    
+    current_xy = (child_node.layout.x, child_node.layout.y)
+
+    while host_node.layout.row != stop_row and host_node.parent_node:
+        parent_node = host_node.parent_node
+        if parent_node.layout.row < host_node.layout.row:
+            v_track = parent_node.iter_track(Track.UPPER_VERTICAL)
+        else:
+            v_track = parent_node.iter_track(Track.LOWER_VERTICAL)
+            while v_track < parent_node.layout.upper_v_track:
+                v_track = parent_node.iter_track(Track.LOWER_VERTICAL)
+        h_track = parent_node.iter_track(Track.HORIZONTAL)
+        offset = parent_node.layout.offset
+
+        sub_parent_xy = (parent_node.layout.x - (offset * v_track), \
+            parent_node.layout.y + (offset * h_track))
+
+        render_loss_branch(sub_parent_xy, current_xy, fig)
+
+        host_node = parent_node
+        current_xy = sub_parent_xy
+    
+    node_xy = (node.layout.x, node.layout.y)
+    mid_xy = (node_xy[0], current_xy[1])
+
+    fig.line(node_xy, mid_xy, PARASITE_EDGE_COLOR)
+    fig.line(mid_xy, current_xy, PARASITE_EDGE_COLOR)
+
+def event_color_shape(event):
+    """
+    Gives the color and shape for drawing event, depending on event type
+    :param event: Event object
+    :return A tuple with the color and shape of an event
+    """
     if event.event_type is EventType.TIPTIP:
-        return LEAF_NODE_COLOR
+        return LEAF_NODE_COLOR, LEAF_NODE_SHAPE
     if event.event_type is EventType.COSPECIATION:
-        return COSPECIATION_NODE_COLOR
+        return COSPECIATION_NODE_COLOR, COSPECIATION_NODE_SHAPE
     if event.event_type is EventType.DUPLICATION:
-        return DUPLICATION_NODE_COLOR
+        return DUPLICATION_NODE_COLOR, DUPLICATION_NODE_SHAPE
     if event.event_type is EventType.TRANSFER:
-        return TRANSFER_NODE_COLOR
+        return TRANSFER_NODE_COLOR, TRANSFER_NODE_SHAPE
     return plot_tools.BLACK
 
-
-def set_host_node_layout(host_tree: tree.Tree):
+def set_host_node_layout(host_tree):
     """
     Sets the logicalRow and logicalCol values of each Node in host_tree.
     Assumes that each host Node has its order set already and this function
     uses those values and structure of the tree to set the logicalRow and logicalCol
     :param host_tree:  A Tree object representing the host tree
     :return None
     """
-    # sets logical row values for leaves in the order they appear in the list of host tree leaves
+    #sets logical row values for leaves in the order they appear in the list of host tree leaves
     logical_row_counter = 0
-    for leaf in host_tree.leaf_list():
+    for leaf in host_tree.leaf_list:
```

Comment:
Please use `leaf_list()` with parenthesis and revert the changes to the class. Generally, we use an instance method as a property (without `()`) only when it is not doing a lot of things. `leaf_list` is iterating through the whole tree, which to me is a lot of things.